lib/rspec/mocks/method_reference.rb

```diff
@@ -94,7 +94,24 @@ def self.method_visibility_for(object, method_name)
           # write `method_missing` in such a way that it handles a dynamic message
           # with private or protected visibility. Ruby doesn't provide you with
           # the caller info.
-          return :public if vis.nil? && object.respond_to?(method_name, false)
+          return :public if vis.nil? && object_responds_to?(object, method_name, false)
+        end
+      end
+
+      # @api private
+      KERNEL_RESPOND_TO_METHOD = ::Kernel.instance_method(:respond_to?)
+
+      if Support::RubyFeatures.supports_rebinding_module_methods?
+        def self.object_responds_to?(object, method_name, visibility)
+          KERNEL_RESPOND_TO_METHOD.bind(object).call(method_name, visibility)
```

Comment:
I guess the solution here is to use `RSpec::Support.method_handle_for` to store respond_to in the proxy at instantiation time (before any stub/mock gets applied). The reason to do this at instantiation time is to remove any kind of conditional logic down the pipe. If we always have the original implementation that'll be nice. Also, proxy instantiation time feels right to me because we know at that point there are no existing stubs/mocks on the object. Now to insert a serious question: do we also need to ensure this doesn't happen via any instance as well?